site/multiDim/MultiDimDataPageContent.tsx

```diff
@@ -132,144 +132,148 @@ export type MultiDimDataPageContentProps = {
+                    const config = {
+                        ...grapherConfig.value,
+                        ...baseGrapherConfig,
                     }
-                })
-                .catch(console.error)
-        },
-        [config, isPreviewing]
-    )
+                    // Batch the grapher updates to avoid getting intermediate
+                    // grapherChangedParams values, which make the URL update
+                    // multiple times while flashing.
+                    // https://stackoverflow.com/a/48610973/9846837
+                    unstable_batchedUpdates(() => {
+                        grapher.setAuthoredVersion(config)
+                        grapher.reset()
+                        grapher.updateFromObject(config)
+                        grapher.downloadData()
+                        grapher.populateFromQueryParams(newGrapherParams)
+                    })
+                }
+            })
+            .catch(Sentry.captureException)
+        return () => {
+            ignoreFetchedData = true
+        }
+    }, [config, isPreviewing, settings, setSearchParams])
 
     const handleSettingsChange = useCallback(
         (settings: MultiDimDimensionChoices) => {
-            const grapher = grapherRef.current
-            if (!grapher) return
-
-            const oldGrapherParams = grapher.changedParams
-            const newGrapherParams: GrapherQueryParams = {
-                ...oldGrapherParams,
-                // Pass the previous tab to grapher, but don't set it in URL. We
-                // want it set only when it's not the default, which is handled
-                // by effect that depends on `grapherChangedParams`.
-                tab: grapher.mapGrapherTabToQueryParam(grapher.activeTab),
-            }
-
-            const { selectedChoices } =
-                config.filterToAvailableChoices(settings)
-
-            setSearchParams(
-                { ...oldGrapherParams, ...selectedChoices },
-                { replace: true }
+            setSettings(
+                config.filterToAvailableChoices(settings).selectedChoices
             )
-            updateGrapher(grapher, selectedChoices, newGrapherParams)
         },
-        [config, setSearchParams, updateGrapher]
+        [config]
     )
 
-    // Set state from query params on page load.
-    useEffect(() => {
-        const grapher = grapherRef.current
-        if (!grapher) return
-        const queryParams = Object.fromEntries(searchParams.entries())
-        updateGrapher(grapher, settings, queryParams)
-        // NOTE (Martin): This is the only way I was able to set the initial
-        // state on page load. Reconsider after the Grapher state refactor, i.e.
-        // when we decouple Grapher state from the Grapher components. Adding
-        // deps properly to the dep array leads to an infinite loop.
-        // eslint-disable-next-line react-hooks/exhaustive-deps
-    }, [])
-
```

Comment:
I was able to get rid of this hack after I realized that duplicating the `settings` state in `useState` and URL params is better that doing this.

Reply:
Welp. I missed an edge case, this won't work. Keeping the hack for now.